lib/streamlit/elements/lib/dialog.py

```diff
@@ -0,0 +1,149 @@
+            )
+        script_run_ctx.has_dialog_opened = True
+
+
+class Dialog(DeltaGenerator):
+    @staticmethod
+    def _create(
+        parent: DeltaGenerator,
+        title: str,
+        *,
+        dismissible: bool = True,
+        width: DialogWidth = "small",
+    ) -> Dialog:
+        block_proto = BlockProto()
+        block_proto.dialog.title = title
+        block_proto.dialog.dismissible = dismissible
+        block_proto.dialog.width = _process_dialog_width_input(width)
+
+        # We store the delta path here, because in _update we enqueue a new proto message to update the
+        # open status. Without this, the dialog content is gone when the _update message is sent
+        delta_path: list[int] = (
+            parent._active_dg._cursor.delta_path if parent._active_dg._cursor else []
+        )
+        dialog = cast(Dialog, parent._block(block_proto=block_proto, dg_type=Dialog))
+
+        dialog._delta_path = delta_path
+        dialog._current_proto = block_proto
+        # We add a sleep here to give the web app time to react to the update. Otherwise,
+        #  we might run into issues where the dialog cannot be opened again after closing
+        time.sleep(0.05)
+        return dialog
+
+    def __init__(
+        self,
+        root_container: int | None,
+        cursor: Cursor | None,
+        parent: DeltaGenerator | None,
+        block_type: str | None,
+    ):
+        super().__init__(root_container, cursor, parent, block_type)
+
+        # Initialized in `_create()`:
+        self._current_proto: BlockProto | None = None
+        self._delta_path: list[int] | None = None
+
+    def _update(self, should_open: bool):
+        """Send an updated proto message to indicate the open-status for the dialog."""
+
+        assert self._current_proto is not None, "Dialog not correctly initialized!"
+        assert self._delta_path is not None, "Dialog not correctly initialized!"
+        _assert_first_dialog_to_be_opened(should_open)
+        msg = ForwardMsg()
+        msg.metadata.delta_path[:] = self._delta_path
+        msg.delta.add_block.CopyFrom(self._current_proto)
+        msg.delta.add_block.dialog.is_open = should_open
+
+        self._current_proto = msg.delta.add_block
+
+        # We add a sleep here to give the web app time to react to the update. Otherwise,
+        #  we might run into issues where the dialog cannot be opened again after closing
+        time.sleep(0.05)
+        _enqueue_message(msg)
+
+    def open(self) -> None:
+        self._update(True)
+
+    def close(self) -> None:
+        self._update(False)
+
+    def __enter__(self) -> Dialog:  # type: ignore[override]
+        # This is a little dubious: we're returning a different type than
+        # our superclass' `__enter__` function. Maybe DeltaGenerator.__enter__
+        # should always return `self`?
+        # self.close()
```

Comment:
Is this an underlying question?

Reply:
yeah I think it is. Its copied from [mutable_status_container](https://github.com/streamlit/streamlit/blame/55bba786dcc8dd85e724746cd26bba3018a147bc/lib/streamlit/elements/lib/mutable_status_container.py#L153-L155), so maybe @LukasMasuch has something to add here.
I have just filed a tech debt item for us to keep track of the question. I think this plays into our notion of "there is room for refactoring on the delta generator side"